File: cpp/questions/diameterOfBinaryTree.cpp
#include <bits/stdc++.h>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        int ans = 0;
        std::function<int(TreeNode*)> dfs = [&ans, &dfs](TreeNode* n) -> int {
            if (n == nullptr) {
                return 0;
            }
            int ld = dfs(n->left);
            int rd = dfs(n->right);
            ans = std::max(ans, ld+rd);
            return std::max(ld, rd)+1;
        };
        dfs(root);
        return ans;
    }
};

File: python/questions/diameterOfBinaryTree.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        ans = 0
        def dfs(n: Optional[TreeNode]) -> int:
            nonlocal ans
            if n is None:
                return 0
            ld = dfs(n.left)
            rd = dfs(n.right)
            ans = max(ans, ld+rd)
            return max(ld, rd)+1
        dfs(root)
        return ans
